Updated pwa content (#579)
MicrosoftDocs/edge-developer: microsoft-edge/progressive-web-apps-edgehtml/get-started.md

    }
    ```  
    

>     If this were a real PWA, you should customize at least the *name*, *start_url*, *short_name*, and *description*, as well as the *icons* \(icons are covered in the next step\).  

Reviewer: ```suggestion
    If this were a real PWA, you would customize *name*, *start_url*, *short_name*, *description*, and *icons* \(icons are covered in the next step\).  
```

    
    To learn more about the different member values and associated purposes, see the [Web App Manifest][MDNWebAppManifest] reference.  
    

> 1.  Next, fill in the empty `icons` array with actual image paths.  For that, use the PWA Builder App Image Generator.  

Reviewer: ```suggestion
1.  Next, fill in the empty `icons` array with actual image paths by using the PWA Builder App Image Generator.  
```

    
    1.  Using a web browser, download this [sample 512x512 PWA image][ImagePwa].  

>     1.  Go to the PWA Builder [App Image Generator][PwaBuilderAppImageGenerator], and select the `pwa.png` image you just saved as the **Input Image** and then click the **Download** button.  

Reviewer: ```suggestion
    1.  Go to the PWA Builder [App Image Generator][PwaBuilderAppImageGenerator], and select the `pwa.png` image you just saved as the **Input Image** and then choose the **Download** button.  
```

    1.  Open \(with Visual Studio or any editor\) the `icons.json` file from the extracted zip and copy the `"icons": [...]` array into the `manifest.json` file of your project.  

1.  Now you must associate your web app manifest with your app.  Open the `layout.pug` file \(in *views* folder\) for editing, and add this line right after the stylesheet link.  \(It is simply the Node [pug][PugAttributes] template shorthand for `<link rel='manifest' href='/manifest.json'>`\).  
    
    ```html
    link(rel='manifest', href='/manifest.json')
    ```  
    
With all that in place, your web app is now serving up a manifest and homescreen-ready app icons!  Try running your app \(`F5`\) and loading up the manifest.  

![Web app manifest loading from localhost][ImageVsNodejsExpressManifest]  

And one of your icons.  

![Square71x71Logo app logo loading from localhost][ImageVsNodejsExpressIcon]  

> If you publish the app live \(with an actual `start_url`\), the Bing search engine now identifies it as a candidate for [automatic packaging and submission to the Microsoft Store][PwaEdgehtmlMicrosoftStore] as an installable Windows 10 app.  Just make sure that your manifest.json file includes the [Quality signals for Progressive Web Apps][WindowsBlogsPwaEdge] that Bing is scanning for the following.  

Reviewer: ```suggestion
If you publish the app live \(with an actual `start_url`\), the Bing search engine now identifies it as a candidate for [automatic packaging and submission to the Microsoft Store][PwaEdgehtmlMicrosoftStore] as an installable Windows 10 app.  Ensure that your manifest.json file includes the [Quality signals for Progressive Web Apps][WindowsBlogsPwaEdge] that Bing scans for as follows: 
```

*   `name`  
*   `description`  
*   At least one icon 512px square or larger \(to ensure an image source of sufficient resolution for auto-generating the splash screen of your app, store listing, tile image, and so on\).  

> In addition, it should be [HTTPS](#https), use [service workers](#service-workers), comply with [Microsoft Store Policies][LegalWindowsAgrementsMicrosoftStorePolicies].  

Reviewer: ```suggestion
In addition, use [HTTPS](#https), [service workers](#service-workers), and comply with [Microsoft Store Policies][LegalWindowsAgrementsMicrosoftStorePolicies].  
```

### HTTPS  

[Service Workers][MDNServiceWorkerApi] and other key PWA technologies that work with service workers \(such as the [Cache][MDNCache], [Push][MDNPushApi], and [Background Sync][MDNSyncManager] APIs\) only work across secure connections, which means [HTTPS][WikiHttps] for live sites or `localhost` for debugging purposes.  

If you [publish this web app as a live site][VisualStudioNodejsTutorialPublishAzureAppService] \(for example, by setting up an [Azure free account][AzureCreateFreeAccount]\), you must ensure your server is configured for HTTPS.  If you use the [Microsoft Azure App Service][AzureWebApps] to host your site, it is served over HTTPS by default.  

> For this guide, you continue using `http://localhost` as a placeholder for a live site served over `https://`.  

Reviewer: ```suggestion
For this guide, continue using `http://localhost` as a placeholder for a live site served over `https://`.  
```

### Service Workers  

> Service Workers is the key technology behind PWAs and act as a proxy between your PWA and the network, enabling your website to act as an installed native app including serving up offline scenarios, responding to server push notifications, and running background tasks.  Service workers also open up all kinds of new performance strategies.  You are not required to implement a full-blown web app to take advantage of the service worker cache for fine-tuned page load performance for your website.  

Reviewer: ```suggestion
Service Workers is the key technology behind PWAs. They act as a proxy between your PWA and the network to enable your website to function as an installed native app that can serve up offline scenarios, respond to server push notifications, and run background tasks.  Service workers also open up new performance strategies.  You are not required to implement a full web app to use the service worker cache for fine-tuned page load performance for your website.  
```

Service workers are event-driven background threads that run from JavaScript files served up alongside the regular scripts that power your web app.  Because Service workers do not run on the main UI thread, service workers do not have DOM access, though the [UI thread][MDNWorkerPrototypePostMessage] and a [worker thread][MDNDedicatedWorkerGlobalScopePostMessage] is able to communicate using `postMessage()` and `onmessage` event handlers.  

You associate a service worker with your app by registering it to the URL origin of your site \(or a specified path within it\).  Once registered, the service worker file is then downloaded, installed, and activated on the user machine.  For more, *MDN web docs* has a comprehensive guide on [Using Service Workers][MDNUsingServiceWorkers] and a detailed [Service Worker API][MDNServiceWorkerApi] reference.  

> For this tutorial, use a ready-made Offline page service worker script courtesy of [PWA Builder][PwaBuilderServiceWorker].  From this, you should customize it with more elaborate functionality according to your needs for performance, network bandwidth, and so on.  Check out the [Service Worker Cookbook][ServiceWorkerCookbook]  provided by Mozilla for a number of useful service worker caching-recipe ideas.  

Reviewer: ```suggestion
For this tutorial, use the Offline page service worker script on [PWA Builder][PwaBuilderServiceWorker].  Start by customizing the script with more functionality according to your performance requirements, network bandwidth, and so on.  Check out the [Service Worker Cookbook][ServiceWorkerCookbook]  provided by Mozilla for a number of useful service worker caching ideas.  
```

1.  Open [https://www.pwabuilder.com/serviceworker][PwaBuilderServiceWorker] and select the \(default\) **Offline page** service worker and click the **Download service worker** button.  
1.  Open the download folder and copy the following two files  

    *   ServiceWorker1\pwabuilder-sw-register.js  
    *   ServiceWorker1\pwabuilder-sw.js  
    
    Save the files to the `public` folder of your Visual Studio web app project.  \(From Visual Studio, use `Ctrl`+`O` to open file explorer to your project and navigate to the `public` folder\).  
    

>     it is worth reviewing the code in both of these files, to get the gist of how to register a service worker that caches a designated page \(`offline.html`\) and serves it when a network fetch fails.  Next, create a simple `offline.html` page as a placeholder for the offline functionality of your app.  

Reviewer: ```suggestion
    It is worth reviewing the code in both of these files, to get the gist of how to register a service worker that caches a designated page \(`offline.html`\) and serves it when a network fetch fails.  Next, create a simple `offline.html` page as a placeholder for the offline functionality of your app.  
```

    ```  
    
    At this point, your `public` folder should have three new files.  
    
    ![Files added to the public folder of the solution][ImageVsNodejsExpressPublic]  
    
1.  In Solution Explorer, open the `routes\index.js` file, and add the following code just before the final command \(`module.exports = router;`\).  
    
    ```javascript
    router.get('/offline.html', function (req, res) {
        res.sendFile('public/offline.html');
    });
    ```  
    
    This instructs your app to serve the `offline.html` file \(when your service worker fetches it for the offline cache\).  
    
1.  Test out your PWA!  Build \(`Ctrl`+`Shift`+`B`\) and Run \(`F5`\) your web app to launch Microsoft Edge and open your `localhost` page.  Then complete the following steps.  
    
    1.  Open the Edge DevTools **Console** \(`Ctrl`+`Shift`+`J`\) and verify the Service worker was registered.  
    1.  In the **Debugger** panel, expand the **Service Workers** control and click on your origin.  In the **Service Worker Overview**, verify your service worker is activated and running.  
        
        ![Edge DevTools Service Worker overview][ImageDevtoolsSwOverview]  
        

>     1.  Still in the Debugger, expand the **Cache** control and verify that the `offline.html` page has been cached.  

Reviewer: ```suggestion
    1.  Expand the **Cache** control and verify that the `offline.html` page is cached.  
```

    }).then(function (clientList) {
        for (var i = 0; i < clientList.length; i++) {
            var client = clientList[i];
            if (client.url == 'http://localhost:1337/' && 'focus' in client)
                return client.focus();
        }
        if (clients.openWindow)
            return clients.openWindow('/');
    }));
});
```  

### Step 6 - Try it out  

Time to test push notifications in your PWA!  

1.  Run \(`F5`\) your PWA in the browser.  Because you modified the service worker code \(`pwabuilder-sw.js`\), you must open the DevTools Debugger \(`F12`\) to the **Service Worker Overview** panel and **Unregister** the service worker and reload \(`F5`\) the page to re-register it \(or you click **Update**\).  In a production scenario, the browser checks regularly check for service worker updates and install the updates in the background.  You should force it here for immediate results.  
    
    As your service worker activates and attempts to subscribe your PWA to push notifications, you should see a permission dialog at the bottom of the page.  
    
    ![Permission dialog for enabling notifications][ImageNotificationPermission]  
    

>     Click **Yes** to enable toast notifications for your PWA.  

Reviewer: ```suggestion
    Choose **Yes** to enable toast notifications for your PWA.  
```

> 1.  From the Service Worker Overview pane, try clicking the  **Push** button.  A toast notification with the \(hard-coded "Test push message from DevTools"\) payload should appear.  

Reviewer: ```suggestion
1.  From the Service Worker Overview pane, try choosing the  **Push** button.  A toast notification with the \(hard-coded "Test push message from DevTools"\) payload should appear.  
```

    
    ![Push a notification from DevTools][ImageDevtoolsPush]  
    

> 1.  Next try clicking the **Send Notification** button on the homepage of your PWA.  This time a toast with the payload from your server appears.  

Reviewer: ```suggestion
1.  Next try choosing the **Send Notification** button on the homepage of your PWA.  This time a toast with the payload from your server appears.  
```

Check out the other PWA guides to learn how to increase customer engagement and provide a more seamless, OS-integrated app experience.  

*   [Windows tailoring][PwaEdgehtmlWindowsFeatures]. Using simple feature detection, you may progressively enhance your PWA for Windows 10 customers through native Windows Runtime \(WinRT\) APIs, such as those for customizing Windows **Start** menu tile notifications and taskbar jumplists, and \(upon permission\) working with user resources, such as photos, music, and calendar.  
*   [PWAs in the Microsoft Store][PwaEdgehtmlMicrosoftStore].  Learn more about the benefits of app store distribution and how to submit your PWA.  

Also be sure to explore the great resources from around the dev community!  [MDN web docs][MDNProgressiveWebApps] and [Google Developers][WebDevProgressiveWebApps] have excellent guides on Progressive Web Apps.  [https://pwa.rocks][ProgressiveWebApps] showcases real-world examples of PWAs, and [https://hnpwa.com][HackerNewsProgressiveWebApps] compares different frameworks and performance patterns for implementing a sample \(Hacker News reader\) PWA.  

<!-- image links -->  

> 

Reviewer: Can we include a See Also section with related links?

Author: It is possible. 
I will start a thread.

Author: See also implemented.
